### PR TITLE
Issue 393 Notification Pattern

### PR DESCRIPTION
**Pull request by NatalieSty**
Add Notification pattern (issue pattern #393）
A Notification is an object that collects information about errors during validation of data. When an error appears the Notification is sent back to the view to display further information about the errors.

Pull request description

-  included Junit tests with 100% coverage
-  ReadMe.md, UML diagram included

### CODE REVIEW

File: notification/README.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
+---
+layout: pattern
+title: Notification
+folder: notification
+permalink: /patterns/notification/
+categories: Behavioural
+language: en
+tags:
+- Notification
```
Comment (iluwatar) [inline notification/README.md:9]: For tags, I would put `Decoupling` and `Presentation`

File: notification/pom.xml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        <version>1.26.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>notification</artifactId>
+
+    <properties>
+        <maven.compiler.source>17</maven.compiler.source>
+        <maven.compiler.target>17</maven.compiler.target>
+    </properties>
```
Comment (iluwatar) [inline notification/pom.xml:17]: I don't think these are needed. We use Java 11 language level for the whole project.

File: notification/pom.xml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-params</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
```
Comment (iluwatar) [inline notification/pom.xml:29]: Please add `maven-assembly-plugin` so that we are able to execute the example from the command line. See the other patterns for examples.

File: notification/src/main/java/com/iluwatar/notification/App.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package com.iluwatar.notification;
+
+/**
+* With the DAO pattern, we can use various method calls to
+* retrieve/add/delete/update data without directly interacting
+* with the data source. The below example demonstrates basic CRUD
+* operations: select, add, update, and delete.
+*/
```
Comment (iluwatar) [inline notification/src/main/java/com/iluwatar/notification/App.java:8]: The text seems to be copied from another pattern. Please update it to introduce the notification pattern briefly and describe how the example code implements it.

File: notification/src/main/java/com/iluwatar/notification/App.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+public final class App {
+
+  /**
+   * This is not called.
+   */
```
Comment (iluwatar) [inline notification/src/main/java/com/iluwatar/notification/App.java:13]: Instead, please describe the class

File: notification/src/main/java/com/iluwatar/notification/App.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+   * Program entry point.
+   *
+   * @param args command line args
+   */
+  public static void main(final String[] args) {
+    final String courseId = "CSE427";
+    final String semester = "Fall21";
+    final String department = "Engineering";
+    final FormRegisterCourse form = new FormRegisterCourse(courseId, semester, department);
+    form.submit();
```
Comment (iluwatar) [inline notification/src/main/java/com/iluwatar/notification/App.java:27]: It's possible to use local variable type inference (keyword `var`) here. Additionally, it would be nice to have some comments that describe what happens. This is an example code for others to read, after all.

File: notification/src/main/java/com/iluwatar/notification/DataTransferObject.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  public Notification notification = new Notification();
+
+  /**
+  * Get a Notification object.
+  *
+  * @return Notification object.
+  */
+  public Notification getNotification() {
+    return notification;
+  }
```
Comment (iluwatar) [inline notification/src/main/java/com/iluwatar/notification/DataTransferObject.java:15]: Use Lombok to get rid of boilerplate code

Comment (iluwatar) [reply notification/src/main/java/com/iluwatar/notification/DataTransferObject.java:15]: Please check this throughout the code

File: notification/src/main/java/com/iluwatar/notification/Error.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package com.iluwatar.notification;
+
+public class Error {
```
Comment (iluwatar) [inline notification/src/main/java/com/iluwatar/notification/Error.java:3]: It's good practice to provide Javadoc for each class

File: notification/src/main/java/com/iluwatar/notification/FormRegisterCourse.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
+package com.iluwatar.notification;
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class FormRegisterCourse {
```
Comment (iluwatar) [inline notification/src/main/java/com/iluwatar/notification/FormRegisterCourse.java:6]: I find the name a little strange. Wouldn't `RegisterCourseForm` sound better?

File: notification/src/main/java/com/iluwatar/notification/Error.java (hunk 8 above)
Comment (iluwatar) [reply notification/src/main/java/com/iluwatar/notification/Error.java:3]: `Error` as a name can be a little bit too generic. Please consider alternatives.

File: notification/src/test/java/com/iluwatar/notification/CourseServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    @Test
+    void testCourseServiceTest() {
+        RegisterCourseDto registerDTO = new RegisterCourseDto();
+        registerDTO.setDepartment("English");
+        registerDTO.setCourseId("CS444");
+        registerDTO.setSemester("Fall21");
+        CourseService course = new CourseService();
+        Boolean isRegistered = course.registerCourse(registerDTO);
+        assertTrue(isRegistered);
+    }
```
Comment (iluwatar) [inline notification/src/test/java/com/iluwatar/notification/CourseServiceTest.java:19]: This probably tests that the happy scenario for registering for a course succeeds. The naming of the test could be better. Additionally, there should be more tests that check the negative cases (course registration does not succeed).

**Review by iluwatar — CHANGES_REQUESTED** [review]

**Comment by NatalieSty** [comment]
@iluwatar thanks for the comments. I have made some changes and committed the change

File: abstract-factory/README.md [hunk 11]
```diff
@@ -210,7 +210,7 @@ Example use cases
 
 ## Consequences
 
-* Dependency injection in java hides the service class dependencies that can lead to runtime errors that would have been caught at compile time.
+* Dependency injection in java hides the service class dependencies that can lead to runtime notificationErrors that would have been caught at compile time.
```
Comment (iluwatar) [inline abstract-factory/README.md:213]: It seems that some text has been accidentally replaced in many other patterns. Please revert the changes.

File: notification/README.md (hunk 1 above)
Comment (iluwatar) [reply notification/README.md:9]: Please remove the tag `Notification`

File: notification/README.md [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+- Decoupling
+- Presentation
+---
+
+## Intent
+
+A Notification is an object that collects information about notificationErrors during validation of data. When an notificationError appears the Notification is sent back to the view to display further information about the notificationErrors.
```
Comment (iluwatar) [inline notification/README.md:16]:
```suggestion
Notification is an object that collects information about notification errors during the validation of data. When a notification error appears, the notification is sent back to the view to display further information about the notification errors.
```

File: notification/README.md [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+
+
+## Explanation
+
+Real world example
+
+> There's a form to register for a course. We need to validate the inputs to make sure 
+> the inputs are valid.
```
Comment (iluwatar) [inline notification/README.md:24]: Please use passive voice (see https://www.grammarly.com/blog/passive-voice/)

File: notification/README.md [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+
+In plain words
+
+> An object that collects together information about notificationErrors and other information and communicates it to the presentation.
```
Comment (iluwatar) [inline notification/README.md:28]:
```suggestion
> Notification is an object that collects together information about notification errors and communicates it to the presentation layer.
```

**Review by iluwatar — CHANGES_REQUESTED** [review]
